transport-native-epoll/src/main/java/io/netty/channel/epoll/EpollRecvByteAllocatorHandle.java

```diff
@@ -119,12 +120,20 @@ public final void readComplete() {
 
     @Override
     public final boolean continueReading(UncheckedBooleanSupplier maybeMoreDataSupplier) {
-        return delegate.continueReading(maybeMoreDataSupplier);
+        return (readPending && maybeMoreDataToRead()) || delegate.continueReading(maybeMoreDataSupplier);
```

Comment:
this will skip the criteria enforced by the delegate (e.g. [MaxMessageHandle](https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/DefaultMaxMessagesRecvByteBufAllocator.java#L142-L145)). The delegate may not respect "maybe more data" and may have other criteria to determine it doesn't want to continue reading. The delegate also checks "auto read" which seems like a more natural fit for the "read pending" check (e.g. `is auto read enabled || someone asked to read`).

Reply:
@Scottmitch hmm true... Let me think about how we could pass it down to the right level.